10 Hoist Module

austingardner opened the pull request:
## Description

This module allows the developer to hoist up an element (usually a modal of some kind) directly under the document.body.
This allows the developer to more easily take care of traps for accessibility. 

## Screenshots
### Before
![Screen Shot 2021-06-02 at 4 02 08 PM](https://user-images.githubusercontent.com/25092249/120562882-2c557c80-c3bc-11eb-9c64-e0f24bde5c67.png)

### After Modal
![Screen Shot 2021-06-02 at 4 02 24 PM](https://user-images.githubusercontent.com/25092249/120562893-324b5d80-c3bc-11eb-9295-e2793647c434.png)

### After unModal
![Screen Shot 2021-06-02 at 4 02 38 PM](https://user-images.githubusercontent.com/25092249/120562901-35dee480-c3bc-11eb-9707-f21df780b550.png)

ianmcburnie commented:
API is looking good! Couple of quick changes/additions:

   * Let's call it `makeup-hoist-element`
   * Please add a README and show an example of the markup before and after an element is hoisted

ianmcburnie commented on docs/makeup-modal/index.html:
```diff
@@ -14,7 +14,8 @@
         <main role="main">
             <h1>makeup-modal demo</h1>
-            <p>The dotted region below has been set to modal, using the modal() method.</p>
+            <p>The dotted region below will be set to modal on click of the "Modal" button, using the modal() method.</p>
```
Is there a reason for changing this demo page, rather than just adding a specific demo page for hoist element? If the API is working then this particular page shouldn't require any changes.

ianmcburnie replied:
Also, for when creating a demo page for hoist-element, there is no need to mention anything about dialogs or traps. The demo should simply show an element being hoisted (probably executed after a button click). The idea being that hoist-element might also have use-cases outside of a dialog scenario.

ianmcburnie replied:
@austingardner Can you revert the changes to the modal demo? The changes have broken the original demo I had. Once reverted we can take a look and figure out a way to do a hoisted demo and a non-hoisted.

ianmcburnie commented on packages/makeup-modal/index.js:
```diff
@@ -23,6 +26,7 @@ function unmodal() {
 function modal(el, options) {
   unmodal();
   modalEl = el;
+  makeupHoist.hoist(modalEl);
```
We should make this configurable (i.e passed in via options). Default should be false for backwards compatibility.

ianmcburnie commented on packages/makeup-hoist-element/package.json:
```diff
@@ -0,0 +1,24 @@
+{
+    "name": "makeup-hoist-element",
+    "description": "Hoists element below body so that traps are more performant, then returns element to original place when closing",
+    "version": "0.1.0",
```
Let's start with 0.0.1 (unless you already did publish it to NPM).

ianmcburnie commented on packages/makeup-modal/index.js:
```diff
@@ -4,25 +4,41 @@ var keyboardTrap = require('makeup-keyboard-trap');
-    keyboardTrap.untrap(modalEl); // let observers know the keyboard is now trapped
+    keyboardTrap.untrap(modalEl);
+
+    if (options.hoist) {
```
Let's use this pattern for options: https://github.com/makeup/makeup-js/blob/master/packages/makeup-screenreader-trap/src/index.js#L104

Might not be the latest and greatest method, but at least we keep it consistent.

austingardner replied:
Okay I'll use the `Object.assign` syntax

agliga commented on docs/makeup-hoist-element/index.js:
```diff
@@ -0,0 +1,15 @@
+    if (hoisted) {
+        Hoist.unhoist(elementToHoist);
+        hoisted = false;
+        return;
```
Instead of having a return can you do an else? It seems much cleaner that way.

agliga commented on docs/makeup-hoist-element/index.js:
```diff
@@ -0,0 +1,15 @@
+const Hoist = require('../../packages/makeup-hoist-element');
+// const { hoist } = require('../../packages/makeup-modal/node_modules/makeup-hoist-element/src');
+
+let hoisted = false;
```
Would it not be better to have a helper to check if the element is hoisted or not? So that the apps do not have to rely on that? Just a thought
So like `Hoist.isHoisted(elementToHoist)`

austingardner replied:
Are you saying get rid of the boolean and manually check if the element is a direct descendant of document.body? I like that actually

agliga commented on packages/makeup-hoist-element/index.js:
```diff
@@ -0,0 +1,79 @@
+}
+
+function hoist(el) {
+  unhoist();
```
I would actually change this to check if the element is hoisted or not and then if it is not, then don't do anything.

agliga commented on docs/makeup-modal/index.js:
```diff
@@ -1,3 +1,31 @@
+    hoistInput.toggleAttribute('checked');
+});
+
+document.addEventListener('hoist', (e) => {
+    console.log(e);
+});
+
+document.addEventListener('unhoist', (e) => {
+    console.log(e);
+});
```
In curious, what are the point of these events? When would be a valid usecase to use them?

agliga commented on packages/makeup-hoist-element/index.js:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    if (modalPlaceholder) {
+      modalEl.removeAttribute('aria-hidden'); // eslint-disable-next-line eqeqeq
+
+      if (prevAriaHiddenSetting != null) {
```
Two questions:
1. Why disable `eqeqeq`?
2. Is the hoisting component the one that's going to handle `aria-hidden`? I thought there was another makeup module to handle that?

ianmcburnie replied:
This module should not do anything with aria-hidden. Only makeup-screenreader-trap should.

austingardner replied:
I disabled `eqeqeq` because I wanted to check for null or undefined. But this line will get deleted anyways.

agliga commented on packages/makeup-hoist-element/index.js:
```diff
@@ -0,0 +1,79 @@
+  unhoist();
+  modalEl = el;
+  modalPlaceholder = document.createElement('div');
+  modalEl.insertAdjacentElement('beforebegin', modalPlaceholder);
```
Any reason why we don't want to use `insertBefore`? 
I think it's more readable that way
https://developer.mozilla.org/en-US/docs/Web/API/Element/insertAdjacentElement#see_also

austingardner replied:
I'll change that; I haven't played with these APIs much before this. `insertBefore` looks good

agliga commented on packages/makeup-hoist-element/src/index.js:
```diff
@@ -0,0 +1,79 @@
+        if (containerDiv) {
+            const childList = Array.from(containerDiv.childNodes);
+            childList.forEach((child) => {
+                if (child.src === undefined) {
```
Can you just do `if (!child.src)`? I assume  `src=""` would also pass but I think it's fine (I don't like `=== undefined` personally, it can lead to bugs if something is set as null)

agliga commented on packages/makeup-hoist-element/src/index.js:
```diff
@@ -0,0 +1,79 @@
+    modalEl.insertAdjacentElement('beforebegin', modalPlaceholder);
+
+    containerDiv = document.createElement('div');
+    const childList = Array.from(document.body.childNodes);
+    childList.forEach((child, index) => {
```
Since childList is not used anywhere else, you can do this as a 1 line
This also works:
`[...document.body.childNodes].forEach((child, index)`, but I'm fine with the `Array.from` syntax. No need to have it in two lines I think 

Also can you explain what is the point of this method? Why are you checking for `child.src`?

austingardner replied:
Well I was separating out the rest of the document.body child nodes from the <script/> tags. I was worried about unintended side effects of moving the location of the <script/> tags. That's also why I keep track of the body indexes - so that I can put everything back in its original location.

agliga commented on packages/makeup-hoist-element/src/index.js:
```diff
@@ -0,0 +1,79 @@
+let modalEl;
+let modalPlaceholder;
+let containerDiv;
+let prevAriaHiddenSetting;
+let bodyIndexes = [];
```
Can we discuss on the naming of these? It's a bit hard to understand. 
* `modalEl` makes sense.
* `modalPlaceholder` seems okay, this is what remembers the old modal position? Maybe call it something like oldModalPositionEl?
* `containerDiv` again not sure what this is? 
* `prevAriaHiddenSettings` it's fine, but is this the value or a setting? If it's just the value true/false, then call it value not settings. It seems like it means it's a configuration
* `bodyIndexes`, again, not sure what this variable is? Maybe get a better name or comment what the point of this is?

Can we call the element that holds the modal in the body something like: `hoistedModalEl` or something (I'm open)? So we know it's the element holding the modal in the body?

ianmcburnie replied:
hoist-element module should have no knowledge of how it is being used, so let's keep the naming "modal"-free.

agliga reviewed (commented):
Some comments, some nits, some questions. 
Looks great overall though!

ianmcburnie commented on packages/makeup-hoist-element/index.js:
```diff
@@ -0,0 +1,79 @@
+
+      var event = document.createEvent('Event');
+      event.initEvent('unhoist', true, true);
+      modalEl.dispatchEvent(event);
```
I'm not sure there's any need for these hoist and unhoist events...?

austingardner replied:
I'll just drop the events

austingardner commented:
Okay I think I addressed all of the comments here... let me know if I need anything else. 

It seemed like the interaction between modal and hoist-element was distorting things, so I took anything referring to the hoist-element out of the makeup-modal and we can add it back in a later PR when we can also add an example for makeup-modal with hoisting and element.

ianmcburnie commented on packages/makeup-hoist-element/README.md:
```diff
@@ -0,0 +1,39 @@
+hoist.hoist(document.querySelector('el'));
+
+// unhoist element
+hoist.unHoist();
```
So I think you are copying the screenreader-trap API here, which is a singleton (makes sense as there can only be a single trap on screen at any time). Are we also saying that there can only be one hoisted element at at time? Something to think about.

austingardner replied:
Yeah in this case keeping things in order would be fairly complicated if you were hoisting more than one thing at a time. 
I think a singleton pattern makes sense.

ianmcburnie commented on packages/makeup-hoist-element/src/index.js:
```diff
@@ -0,0 +1,70 @@
+module.exports = {
+    hoist,
+    unhoist,
+    isHoisted
```
I don't think there's any reason to export `isHoisted`.

ianmcburnie commented on packages/makeup-hoist-element/src/index.js:
```diff
@@ -0,0 +1,70 @@
+
+    containerDiv = document.createElement('div');
+    [...document.body.childNodes].forEach((child, index) => {
+        if (!child.src) {
```
Is the intention here to filter out script tags? If so, we should create a filter function that filters out array elements based on tagName.

austingardner replied:
Well here's the issue: I needed to keep the script tags relatively in the same place when I `hoist`, and then when I `unhoist` I need to move them back to the same place I started with. That was my objective with this `forEach`.

ianmcburnie commented on packages/makeup-hoist-element/src/index.js:
```diff
@@ -0,0 +1,70 @@
+    if (hoistEl) {
+        if (containerDiv) {
+            [...containerDiv.childNodes].forEach((child) => {
+                if (!child.src) {
```
Possibly no need to check again if these elements already got filtered out during hoist()

ianmcburnie commented on docs/makeup-hoist-element/index.js:
```diff
@@ -0,0 +1,11 @@
+const elementToHoist = document.querySelector('#hoist-this');
+
+elementToHoist.addEventListener('click', function() {
+    if (Hoist.isHoisted(elementToHoist)) {
```
I would remove the `isHoisted` method and just let the app/consumer track state of whether they hoisted the element or not.

ianmcburnie commented on docs/makeup-hoist-element/index.html:
```diff
@@ -0,0 +1,26 @@
+            first child
+            <div>
+                second child
+                <div id="hoist-this">
```
Can you add one or two more hoistable element examples to the page so that we know the singleton pattern is working (i.e. we should see the currently hoisted element get unhoisted). I suggest adding an example of the edge case where an element is already a direct descendant of the body.

ianmcburnie commented:
After chatting offline we have decided to just fold this logic directly into the makeup-modal module. Main reason being we don't foresee any use case for this functionality outside of a modal use case.
